security/advancedtls/advancedtls.go

```diff
@@ -187,39 +205,105 @@ func (o *ClientOptions) config() (*tls.Config, error) {
 		return nil, fmt.Errorf(
 			"client needs to provide custom verification mechanism if choose to skip default verification")
 	}
+	// Make sure users didn't specify more than one fields in
+	// RootCertificateOptions and IdentityCertificateOptions.
+	rootOptionNum := o.RootCertificateOptions.nonNilFieldCount()
+	if rootOptionNum > 1 {
+		return nil, fmt.Errorf(
+			"at most one field in RootCertificateOptions could be specified")
+	}
+	identityOptionNum := o.IdentityCertificateOptions.nonNilFieldCount()
+	if identityOptionNum > 1 {
+		return nil, fmt.Errorf(
+			"at most one field in IdentityCertificateOptions could be specified")
+	}
+	if o.GetIdentityCertificatesForServer != nil {
+		return nil, fmt.Errorf(
+			"GetIdentityCertificatesForServer cannot be specified on the client side")
+	}
+	// Use system default trust certificate if no verification specified.
 	rootCAs := o.RootCACerts
-	if o.VType != SkipVerification && o.RootCACerts == nil && o.GetRootCAs == nil {
+	if o.VType != SkipVerification && rootOptionNum == 0 {
 		// Set rootCAs to system default.
 		systemRootCAs, err := x509.SystemCertPool()
 		if err != nil {
 			return nil, err
 		}
 		rootCAs = systemRootCAs
 	}
-	// We have to set InsecureSkipVerify to true to skip the default checks and
-	// use the verification function we built from buildVerifyFunc.
+	// Fill the root and identity reloading functions based on the Provider
+	// implementation.
+	if o.RootProvider != nil {
```

Comment:
If we simply document the order of preference of the options, and then make sure that at least one of them is set, then we could use a simple switch statement to pick the highest priority option which is set while constructing the `tls.Config`.
The idea is here: https://play.golang.org/p/ru5-mzMbPiz

Reply:
Because of the concerns mentioned above, I think it might be slightly better to keep the check. But using switch does make the code clearer, and I've already updated the code to reflect the change.